### ISSUE TITLE
Embed an inner NLP as a block in an outer model: POUNCE forward solve with sIPOPT sensitivities (1st and 2nd order)

### ISSUE THREAD

**Issue by robertojjx** [jkitchin/discopt#1216]
## Summary

I would like to embed an inner NLP as a block inside an outer model: pass the inner
optimization problem as it is — objective and constraints, no reformulation — have its
forward pass solved by POUNCE, and have POUNCE's existing sensitivity machinery supply the
derivatives the outer solver needs.

Most of the machinery is already present. `discopt.solvers.sipopt.pounce_sensitivity` takes
a `dm.Model` unchanged, solves it with POUNCE, and returns the sIPOPT sensitivity system

```
[W  Jᵀ] [dx*/dp]  =  -[∂²L/∂x∂p]
[J   0] [dλ*/dp]      [∂g/∂p   ]
```

as `SensitivityResult.dx_dp` (n × n_params) and `.dlambda_dp` (m × n_params).

Two gaps prevent using it as an inner block:

1. It returns arrays to Python. There is no construct that makes the solve a node in an
   outer `Model`, so an outer POUNCE solve cannot consume the derivatives.
2. It stops at first order. There is no `d²x*/dp²`. `ReducedHessian` exposes only
   `is_positive_definite`, and `QpSensitivity.reduced_hessian` is QP-only.

## Why the existing workaround is insufficient

Without such a block, embedding a follower means writing its KKT conditions by hand into
`dm.implicit_full_space`. For a non-convex follower this is a relaxation rather than a
reformulation: the KKT set contains every stationary point, and the leader is free to select
whichever one favours its own objective.

The follower below is a projection onto the unit circle. The objective is strictly convex,
but the nonlinear equality makes the feasible set non-convex, so the KKT system has two
roots — the near side (the minimizer, `y* = (p,1)/√(p²+1)`, hence `y0 > 0` for `p > 0`) and
the far side (a maximizer). The leader asks for `y0 = -0.9`, which no projection can produce.

```python
import numpy as np
import discopt.modeling as dm
from discopt.modeling import Model

def kkt_residual(u, v):
    p = u[0]
    y0, y1, lam = v[0], v[1], v[2]
    return [2 * (y0 - p) + 2 * lam * y0,
            2 * (y1 - 1.0) + 2 * lam * y1,
            y0 * y0 + y1 * y1 - 1.0]

m = Model("leader")
p = m.continuous("p", lb=0.05, ub=3.0)
v = dm.implicit_full_space(m, kkt_residual, [p], n_unknowns=3,
                           bounds=(np.array([-2.0, -2.0, -10.0]),
                                   np.array([2.0, 2.0, 10.0])))
m.minimize((v[0] + 0.9) ** 2)
r = m.solve()

pv = float(np.asarray(r.value(p)).ravel()[0])
yv = np.asarray(r.value(v)).ravel()[:2]
true = np.array([pv, 1.0]) / np.sqrt(pv ** 2 + 1.0)
print(r.status, pv, yv, true)
```

```
status         : optimal
p*             : 2.064741
y* returned    : [-0.9     -0.43589]
true projection: [ 0.9      0.43589]
leader objective at returned y* : 1.415e-15
leader objective at true y*     : 3.240e+00
```

The returned `y*` is not the projection of any point; it is the far-side KKT root, reported
as optimal. `dm.implicit` shows the same behaviour and is additionally dependent on the
starting point: from a near-side `x0` it recovers the correct outer optimum
(`p* = 2.064741` against an analytic `2.064742`), and from a far-side `x0` it terminates at
`p* = 0.05`, the lower bound.

Solving the same follower as an optimization problem does not have this failure mode,
because only minimizers are representable. Using `make_nlp_layer`, which wraps the same
POUNCE machinery for JAX:

```
exact  y* = [0.4472136  0.89442719]   dy*/dp = [ 0.71554175 -0.35777088]
layer  y* = [0.4472136  0.89442719]   err = 8.02e-13
layer dy*/dp = [ 0.71554175 -0.35777088]   err = 3.78e-11

p=0.2: y* = [0.196116 0.980581]   minimizer
p=1.0: y* = [0.707107 0.707107]   minimizer
p=2.0: y* = [0.894427 0.447214]   minimizer
```

The correct behaviour is therefore already implemented; it is not reachable from the
algebraic modelling layer.

## Request

**1. An argmin block that takes the inner model unchanged.**

```python
inner = Model("projection")     # objective and constraints, untouched
...
v = dm.argmin(inner, bind={p_inner: outer_expression})
m.minimize(f(v))
```

Forward pass: a POUNCE solve of `inner`. Derivatives passed to the outer solver: the KKT
sensitivity system `pounce_sensitivity` already assembles.

**2. Second-order sensitivities.** `d²x*/dp²` from the same factorized KKT system, so an
outer NLP solver can use exact Hessians instead of a quasi-Newton approximation. First order
alone is already useful, so an `order=1|2` switch would be a reasonable compromise if second
order proves expensive.

**3. Exact right-hand side.** `pounce_sensitivity` currently forms `∂²L/∂x∂p` and `∂g/∂p`
by central finite differences on parameter values. Exact derivatives from the tape would
improve accuracy and cost, and matter more once second order is built on top.

## Relation to a Pyomo grey box

The contract is that of Pyomo's `ExternalGreyBoxModel`, with the forward pass being an
optimizer rather than a simulator.

| | Pyomo `ExternalGreyBoxModel` | requested `dm.argmin` |
|---|---|---|
| what is passed | callback plus hand-written Jacobians | the model, unchanged |
| forward pass | user callback, typically a simulator | POUNCE solve |
| derivatives | user-supplied; Hessian optional | POUNCE sIPOPT machinery, 1st and 2nd order |
| outer solver | cyipopt through PyNumero only | POUNCE |
| `.nl` export | not possible (Python callback) | desirable |

The Pyomo route requires cyipopt and cannot be written to `.nl`, which prevents comparing
solvers on an identical formulation. discopt has neither restriction; the gap is only the
missing connection between an inner solve and the outer model.

## Existing components this would compose

- `solvers/sipopt.pounce_sensitivity` — model as-is, POUNCE solve, `dx_dp` and `dlambda_dp`.
- `_relax/pounce_layer.make_nlp_layer` — the same idea already bridged to JAX through
  `custom_vjp`. Reverse-mode only: `jax.hessian` raises *"can't apply forward-mode autodiff
  (jvp) to a custom_vjp function"*.
- `dm.implicit_full_space` — demonstrates how a block can be lowered so that the Rust tape,
  and therefore POUNCE, handles it.

## Environment

discopt 0.8.0, pounce-solver 0.11.0, Python 3.10, macOS arm64.
